### ISSUE TITLE
ReduceLROnPlateau Callback behaves unexpectedly when cooldown == 0

### ISSUE THREAD

**Issue by mjdietzx** [fchollet/keras#3991]
In this case:

```python
reduce_lr = ReduceLROnPlateau(monitor='val_loss', patience=4, cooldown=0)
model.fit(X_train, Y_train, callbacks=[reduce_lr])
 ```

After 4 epochs of `val_loss` not improving, we reduce the learning rate as expected. However, since `self.cooldown == 0` we don't take this [branch](https://github.com/fchollet/keras/blob/master/keras/callbacks.py#L620) and therefore never reset `self.wait`. Which causes:

1. Learning rate reduced after 4 epochs of no improvement.
2. Learning rate reduced again the next epoch since we don't reset `self.wait`.
3. Every single epoch we are reducing the learning rate after the first time we reduce it.
4. Once `val_loss` improves, we reset `self.wait` and now wait for `self.patience` epochs before reducing LR again.

In the case where `self.cooldown > 0` this code should work as expected.

Gist: https://gist.github.com/mjdietzx/3aaf9c58486f6e6ff310a0d960d8bb4e
@basveeling

**Comment by basveeling** [comment]
Thanks! I'll take a look this weekend and send a new PR.

**Comment by mjdietzx** [comment]
Looks like it'll be an easy fix but if you're tight on time/need any help just let me know.

**Comment by superhans** [comment]
Hi Guys,
I'm not 100% sure that this is working the way we expect it to. Please correct me if I'm wrong : 
My call backs look like this (on the mnist_mlp.py example): 
`reduce_lr = ReduceLROnPlateau(monitor='val_loss', patience=0, cooldown=0, 
 verbose=1, min_lr=0.001)

history = model.fit(X_train, Y_train,
                    batch_size=batch_size,
                    verbose=1, 
                    validation_data=(X_test, Y_test),
                    callbacks=[reduce_lr],
                    nb_epoch=nb_epoch)
`
However, having stepped through the code, the lr is never reduced (the statement never prints even though its in verbose mode), and I put a breakpoint on the line where the old learning rate gets multiplied by the "factor", but it never reaches that line. My val loss during that period looks like this : 

val_loss: 0.1053
                
val_loss: 0.0976
                
val_loss: 0.0717
                
val_loss: 0.0902
                
val_loss: 0.0770
                
val_loss: 0.0712
                
val_loss: 0.0775
                
val_loss: 0.0862
                
val_loss: 0.0780
                
val_loss: 0.0893
                
val_loss: 0.0916

So am I doing something wrong (btw, I tested it on the latest keras downloaded today) ?
